test/common/http/conn_manager_impl_test.cc

```diff
@@ -2835,134 +2835,66 @@ TEST_F(HttpConnectionManagerImplTest, RequestTimeoutIsDisarmedOnConnectionTermin
+
+  Buffer::OwnedImpl first_line("GET /HTTP/1.1\r\n");
+  Buffer::OwnedImpl second_line("Host: localhost:8080\r\n");
+  Buffer::OwnedImpl empty_line("\r\n");
+  conn_manager_->onData(first_line, false);
+  EXPECT_TRUE(request_header_timer->enabled_);
+  conn_manager_->onData(second_line, false);
+  EXPECT_TRUE(request_header_timer->enabled_);
+  conn_manager_->onData(empty_line, false);
+  EXPECT_FALSE(request_header_timer->enabled_);
 
   expectOnDestroy();
+  EXPECT_CALL(*request_header_timer, disableTimer).Times(1);
   filter_callbacks_.connection_.raiseEvent(Network::ConnectionEvent::RemoteClose);
 }
 
 TEST_F(HttpConnectionManagerImplTest, RequestHeaderTimeoutCallbackDisarmsAndReturns408) {
   request_headers_timeout_ = std::chrono::milliseconds(10);
   setup(false, "");
 
-  std::string response_body;
+  Event::MockTimer* request_header_timer;
   EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance&) -> Http::Status {
-    Event::MockTimer* request_header_timer = setUpTimer();
+    request_header_timer = setUpTimer();
     EXPECT_CALL(*request_header_timer, enableTimer(request_headers_timeout_, _)).Times(1);
-    EXPECT_CALL(*request_header_timer, disableTimer()).Times(AtLeast(1));
-
-    EXPECT_CALL(response_encoder_, encodeHeaders(_, false))
-        .WillOnce(Invoke([](const ResponseHeaderMap& headers, bool) -> void {
-          EXPECT_EQ("408", headers.getStatusValue());
-        }));
-    EXPECT_CALL(response_encoder_, encodeData(_, true)).WillOnce(AddBufferToString(&response_body));
 
     conn_manager_->newStream(response_encoder_);
     EXPECT_CALL(filter_callbacks_.connection_.dispatcher_, setTrackedObject(_)).Times(2);
-    request_header_timer->invokeCallback();
-    return Http::okStatus();
-  }));
-
-  Buffer::OwnedImpl fake_input("1234");
-  conn_manager_->onData(fake_input, false); // kick off request
-
-  EXPECT_EQ(1U, stats_.named_.downstream_rq_header_timeout_.value());
-  EXPECT_EQ("request header timeout", response_body);
-}
-
-TEST_F(HttpConnectionManagerImplTest,
-       RequestHeaderTimeoutIsNotDisarmedOnIncompleteRequestWithHeader) {
-  request_headers_timeout_ = std::chrono::milliseconds(10);
-  setup(false, "");
-
-  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance&) -> Http::Status {
-    Event::MockTimer* request_header_timer = setUpTimer();
-    EXPECT_CALL(*request_header_timer, enableTimer(request_headers_timeout_, _)).Times(1);
-    EXPECT_CALL(*request_header_timer, disableTimer()).Times(1);
-
-    RequestDecoder* decoder = &conn_manager_->newStream(response_encoder_);
-    RequestHeaderMapPtr headers{
-        new TestRequestHeaderMapImpl{{":authority", "host"}, {":path", "/"}, {":method", "GET"}}};
-
-    // the second parameter 'false' leaves the stream open
-    decoder->decodeHeaders(std::move(headers), false);
     return Http::okStatus();
   }));
 
-  Buffer::OwnedImpl fake_input("1234");
+  Buffer::OwnedImpl fake_input("GET /resource HTTP/1.1\r\n\r\n");
```

Comment:
This looks like a complete request...  Why isn't the time out disabled?!  Oh, I see - this test mocks out everything.  No actual parsing of request headers is happening and timeouts are created, enabled and disabled from mock methods.

Testing just at this level doesn't feel complete.  It would be good to have some tests that advance simulated time forward and trigger this timeout while in the middle of processing of H1 and H2 request headers.

Reply:
Added in test/integration/http_timeout_integration_test.cc.